### ISSUE TITLE
Proposal: use prelude::*

### ISSUE THREAD

**Issue by ratijas** [woboq/qmetaobject-rs#83]
This is the proposal draft to include a `prelude` module into `qmetaobject` crate, similar to the one in Rust standard library and many other existing crates. The purpose of the module shall be to re-export all or most of public items from the crate.
Questions to consider:
 - [ ] should all items be exported or only subset of them?
 - [ ] if not all, then how to define a subset?

certainly relates to #81

**Comment by ogoffart** [comment]
Sounds like a good idea

**Comment by ratijas** [comment]
How about this?

- [ ] Put all the wrappers in a `qmetaobject::q` sub-module
- [ ] Put each class' wrapper with all its static, non-static members and operators implementation in a separate file, like Qt does.
    For example, let there be `qmetaobject::q::string::QString` struct wrapper.
- [ ] Make all those sub-modules public
- [ ] Re-export their content in prelude
- [ ] Write a migration guide for 2.0, since it's a hell of a breaking change

**Comment by ogoffart** [comment]
I'm not seeing the advantage of the `q` submodule.

**Comment by ratijas** [comment]
Sorry, I have had clarified this. In Qt everything is inside QSomething header files, so we might use qsomething for module names, but it drives linter (spell checker) insane and generally looks bad in lowercase. On the other hand, keeping wrapper stuff in a separate namespace makes it easier to differentiate between them and the custom library runtime like logs or signal/slot connection proxies.

**Comment by ogoffart** [comment]
I'm thinking actually the wrapper could go in their own crate.
I have been wanting to rename this crate anyway because it is not really explanatory.
But i see there is value in having a "qttypes"  or "qtwrapper" crate that contains all the type.  keeping the qmetaobject_derive internal crate as it,  and renaming all the rest in a more generic "qtquick" (nah, already taken) or "qml4rust"  or i don't know.

**Comment by ratijas** [comment]
> the wrapper could go in their own crate

That's definitely an option to consider in the first place. Although, it seems like wrappers and runtime are somewhat cross-dependent, but I wouldn't be myself, if I'd draw to the conclusions without verifying the hypothesis first.

Throwing some random ideas:

 - %cratename%-derive (macros)
 - %cratename%-rt (runtime)
 - %cratename% (actually wrappers)

For me, current name is fine, but that may be a baby duck syndrome. Either this, or exact name as upstream project ("qt"), or something completely unrelated to programming field whatsoever (like `darling` crate).

**Comment by rubdos** [comment]
Chipping in: since a while you expose a `log` module, which means that for people using `use qmetaobject::*` in combination with the fairly standard `log` crate, we're suddenly getting a ton of:

```
error[E0433]: failed to resolve: could not find `trace` in `log`
  --> src/actor/messageactor.rs:78:14
   |
78 |         log::trace!("MessageActor has a registered storage");
   |              ^^^^^ could not find `trace` in `log`
```

I think a `prelude` module would probably the cleanest solution (then we'd not be glob-import qmetaobject anymore, as you currently recommend in the documentation), except if you folks want to possibly rename the `log` module to something like `qlog`.

Let me know if this should (also) be a separate issue!

**Comment by rubdos** [comment]
If having *some* `prelude` is preferred over not having one at all, I just added this to my own code:

```rust
pub mod qmetaobject_prelude {
    pub use qmetaobject::{
        qt_base_class, qt_method, qt_property, qt_signal, QAbstractListModel, QByteArray, QDate,
        QDateTime, QModelIndex, QObject, QObjectBox, QString, QTime, QVariant, __Deref,
        lazy_static, qmetaobject_lazy_static,
    };
}
```

I think that includes most useful stuff already.

(We can now *almost* build Whisperfish on mainline qmetaobject-rs, we're really close :-D)

**Comment by rubdos** [comment]
FYI, we can now build Whisperfish without any modifications to qmetaobject. We're using 0.2 from crates.io  :-D

**Comment by ogoffart** [comment]
>  We're using 0.2 from crates.io

Just published 0.2.1 with contains the prelude
